chia/full_node/mempool_manager.py

```diff
@@ -463,13 +469,27 @@ async def validate_spend_bundle(
                 EligibilityAndAdditions(is_eligible_for_dedup=False, spend_additions=[], ff_puzzle_hash=None),
             )
             mark_as_fast_forward = eligibility_info.ff_puzzle_hash is not None and supports_fast_forward(coin_spend)
+            unspent_lineage_info = None
             if mark_as_fast_forward:
                 # Make sure the fast forward spend still has a version that is
                 # still unspent, because if the singleton has been melted, the
                 # fast forward spend will never become valid.
                 assert eligibility_info.ff_puzzle_hash is not None
-                if await get_unspent_lineage_info_for_puzzle_hash(eligibility_info.ff_puzzle_hash) is None:
-                    return Err.DOUBLE_SPEND, None, []
+                # Have we already seen this singleton?
+                unspent_lineage_info = singleton_latest_unspent_info_map.get(eligibility_info.ff_puzzle_hash)
+                if unspent_lineage_info is None:
+                    # We haven't, so let's look it up
+                    unspent_lineage_info = await get_unspent_lineage_info_for_puzzle_hash(
+                        eligibility_info.ff_puzzle_hash
+                    )
+                    if unspent_lineage_info is None:
+                        # This singleton has been melted
+                        return Err.DOUBLE_SPEND, None, []
+                    singleton_latest_unspent_info_map[eligibility_info.ff_puzzle_hash] = unspent_lineage_info
+                # Keep track of the latest unspent info for this fast forward
+                fast_forward_unspents_map.setdefault(
```

Comment:
I don't see you ever delete entries from this map, just insert. It looks like it will grow indefinitely until you run out of RAM

Reply:
On new blocks, we prepare a new map to replace the old one, and when doing so, we remove the old spent version and track the new one: https://github.com/Chia-Network/chia-blockchain/pull/19331/files#diff-ee823ddd491e7037efb6bda95223e6e4aa703239fd08f28c54a6e3a51b75ff42R874-R877